### ISSUE TITLE
support shothand-syntax

### ISSUE THREAD

**Issue by AndyOGo** [AndyOGo/stylelint-declaration-strict-value#11]
see https://github.com/sh-waqar/stylelint-declaration-use-variable/issues/16

**Comment by ixley** [comment]
any updates on this? I just tried switching to this plugin since sh-waqar/stylelint-declaration-use-variable doesn't support exceptions but immediately noticed that shorthand values cause issues (and [$variables made negative](#13) also don't seem to be supported).

**Comment by jitendravyas** [comment]
+1 I'm also waiting for this feature

**Comment by AndyOGo** [comment]
@jitendravyas 
I'm keen for this feature too. Though it's ambiguous to specify `border` shorthand.
It could mean one of the following:

- A `$valiable` specifying all three properties, like:
  ````scss
  $border: 1px solid #fff;

  .foo { border: $border; }
  ````
- Or multiple `$variables` specifying each property, like:
````scss
  $border-width: 1px:
  $border-style: solid;
  $border-color: #fff;

  .foo { border: $border-width $border-style $border-color; }
  ````
- or any other possible shorthand combination

So in my opinion the spec schema needs to be extend to allow to specify each property check of shorthand props.

**Comment by jrwebdev** [comment]
Not sure if this is related but I'm seeing the following:

```
margin: 0 $margin; // error
margin: $margin 0; // ok
margin: auto $margin; // ok
margin: $margin auto; // ok
margin: auto $margin auto; // ok
margin: auto $margin auto $margin; // ok
margin: $margin 0 $margin 0; // ok
margin: $margin 20px; // ok (shouldn't be)
margin: 20px $margin; // ok (shouldn't be)
```

Config:

```
'scale-unlimited/declaration-strict-value': [
      ['/color/', '/margin/', '/padding/', 'box-shadow', 'z-index'],
      {
        ignoreKeywords: ['currentColor', 'transparent', 'inherit', 'auto', 'none', '0'],
      },
    ],
```
Using an integer value for 0 in the config throws a linting validation error.

**Comment by AndyOGo** [comment]
@jrwebdev 
Thanks for your feedback.
Sure this is related, this feature is about supporting short-hand syntax like your `margin: ...` above.

Regarding your `'0'`, it's not a valid CSS keyword

**Comment by OriR** [comment]
Another use case I found that may be solvable but is still related to the shorthand syntax:
```scss
border-color: transparent transparent $color1 $color2;
```

assuming this config:
```json
"scale-unlimited/declaration-strict-value": [
  ["/color/", "font-size", "font-weight"],
  {
    "ignoreKeywords": ["currentColor", "inherit", "transparent"]
  }
],
```

I would expect it to not flag it as an error since it can understand both `transparent` as an ignored keyword and `$color1`, `$color2` as variables.

**Comment by tvooo** [comment]
Another issue I found is when using a function with the arguments spread over multiple lines. Not sure if this is related or I should create a new issue.

This, for example, will fail:

```
background-color: get-color(
  $arg1,
  $arg2
);
```

**Comment by AndyOGo** [comment]
**CSS Shorthand Syntax expansion** is released with [`v1.4.0`](https://www.npmjs.com/package/stylelint-declaration-strict-value).
Please read more in the [docs](https://github.com/AndyOGo/stylelint-declaration-strict-value#css-shorthand-syntax)

**Comment by mexn** [comment]
Hi,

I might have not got the concept, or there's probably still an issue, but `margin: 0 auto;` is (still) marked as error. I'm using version 1.4.4 and this is my config:

```json
{
  "scale-unlimited/declaration-strict-value": [
    [
      "color",
      "/border-?(color|radius|width)?/",
      "/margin-?(top|right|bottom|left)?/",
      "/padding-(top|right|bottom|left)?/",
      "z-index",
      "line-height",
      "/font-(size|weight)/",
      "/background-?(color|)?/",
      "box-shadow"
    ], {
      "expandShorthand": true,
      "ignoreValues": ["inherit", "initial", "unset", "0", "none", "currentColor", "transparent", "auto"]
    }
  ]
}
```

For the margin-property, I tried several versions, only `margin`, `margin` and all the direction-versions, the regex above, but nothing seems to do the trick. Am I missing something?

**Comment by AndyOGo** [comment]
@mexn your config looks correct to me.

A reduced test-case would be awesome.

**Comment by mexn** [comment]
@AndyOGo I'm not quite sure, what exactly you expect as reduced test-case, but it's basically just
```scss
.class {
      margin: 0 auto;
}
```

Only `0` or `auto` does work, just the combination of both, irregardless of the order, fails.

**Comment by AndyOGo** [comment]
Passt!
Hab grad gesehen du kommst aus Wien, ich bin dort aufgewachsen :)
Danke für dein Feedback.

Ich werde am Abend einen test-case im repo erfassen und einen PR machen.
Alle tests sind hier: https://github.com/AndyOGo/stylelint-declaration-strict-value/tree/master/test

**Comment by mexn** [comment]
😉 Großartig, danke!

**Comment by AndyOGo** [comment]
Thank you for your great feedback @mexn 
I released a fix in [`v1.4.5`](https://www.npmjs.com/package/stylelint-declaration-strict-value/v/1.4.5). Hope that works for you.

**Comment by mexn** [comment]
@AndyOGo Thank you for the fix, this works now! 👍
